### PR TITLE
Add Curvlinops backend & add default `functorch` implementations of many curvature quantities

### PR DESCRIPTION
**Pull request by wiseodd**
Changelog:
1. Add [Curvlinops](https://github.com/f-dangel/curvlinops/tree/main) backend.
2. Add default `functorch` implementation for many quantities: (in `CurvatureInterface` itself)
    1. Jacobians 
    2. Batch gradients
    3. Diag GGN (both stochastic and exact) and diag EF
    4. Full GGN (both stochastic and exact) and full EF

**Rationale for (1):** BackPACK & ASDL are not actively supported anymore and are hard to extend. Note that we use Curvlinops mainly for KFACs. 

**Rationale for (2):** `functorch` is a first-class PyTorch module => has the highest compatibility to any PyTorch models => futureproofing `laplace-torch`, less dependence to third-party backends

**Together:** They make `laplace-torch` very flexible and compatible with most architectures out there (and in the future).

With this PR, `Laplace(backend=CurvlinopsBackend)` supports:
1. Diag-GGN-exact, diag-GGN-MC, diag-EF for both classification and regression
2. Kron-GGN-exact, kron-GGN-MC, kron-EF for both classification and regression
3. Full-GGN-exact, full-GGN-MC, full-EF, full-Hessian-exact for both classification and regression

Note that previous backends only support some subsets of them.

I'd like to ask @runame to specifically review this since he's involved in Curvlinops. Specifically, for the K-FAC backend proposed here.

### CODE REVIEW

File: laplace/curvature/curvlinops.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            separate_weight_and_bias=False
+        )
+        linop._compute_kfac()
+
+        kron = self._get_kron_factors(linop, len(y))
+        kron = self._rescale_kron_factors(kron, len(y), N)
+
+        loss = self.lossfunc(self.model(X), y)
+
+        return self.factor * loss.detach(), self.factor * kron
```
Comment (wiseodd) [inline laplace/curvature/curvlinops.py:74]: @runame please check this implementation. I'm not sure how to proceed. Comparing to BackPACK's & ASDL's KFAC, I still get the wrong results. 

If you need a test script: <https://gist.github.com/wiseodd/d9445133635a5d1d01f5b4c6e387586c>

File: laplace/curvature/curvlinops.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        kron *= self.factor
+
+        # Correction to match BackPACK & ASDL results
+        for factors in kron.kfacs:
+            for i in range(len(factors)):
+                factors[i] /= 2
+
+        loss = self.lossfunc(self.model(X), y)
+
+        return self.factor * loss.detach(), kron
```
Comment (wiseodd) [inline laplace/curvature/curvlinops.py:79]: @runame I found the culprit. Curvlinops' KFAC factors differ are twice as large as BackPACK's and ASDL's. So I put a correction in L72-75 above. 

I'm not sure what's happening. Please propose a more elegant fix.

**Review by runame — COMMENTED** [review]
I reviewed the `curvlinops` part for now and pushed fixes directly.

I noticed that our tests are not systematic enough and that we are currently missing a few important cases (e.g., KFAC(-expand) is not the same with backpack and curvlinops for `Conv2d` modules). Since this branch still uses the old ASDL version, I think we should fix the tests of the curvature backends after this PR and #144 are merged.

File: laplace/curvature/curvlinops.py (hunk 2 above)
Comment (runame) [reply laplace/curvature/curvlinops.py]: Fixed in [this commit](https://github.com/aleximmer/Laplace/pull/146/commits/8d8b8bb4eb90df05efcf79f6b8da8d67ecb29127). The issue was that `linop._compute_kfac()` was called twice, so that all Kronecker factors were added to itself once, resulting in the factor of 2. This can be avoided (also more efficient) by setting `check_deterministic=False`.

**Comment by wiseodd** [comment]
@aleximmer please give a final check and feel free to merge!

File: laplace/baselaplace.py [hunk 3]
```diff
@@ -791,7 +789,7 @@ def _init_H(self):
         self.H = torch.zeros(self.n_params, self.n_params, device=self._device)
 
     def _curv_closure(self, X, y, N):
-        return self.backend.full(X, y, N=N)
+        return self.backend.full(X, y)
```
Comment (aleximmer) [inline laplace/baselaplace.py:792]: Why does this have to be removed? I think there is reason for having it in case of subset-of-data approximations and it might be already used in Asdl. I would be in favor of not changing it unless necessary.

File: laplace/curvature/curvature.py [hunk 4]
```diff
@@ -39,13 +40,18 @@ def __init__(self, model, likelihood, last_layer=False, subnetwork_indices=None)
...
+        self.params_dict = {k: v for k, v in self._model.named_parameters() if v.requires_grad}
+        self.buffers_dict = {k: v for k, v in self.model.named_buffers()}
 
     @property
     def _model(self):
         return self.model.last_layer if self.last_layer else self.model
 
     def jacobians(self, x, enable_backprop=False):
-        """Compute Jacobians \\(\\nabla_\\theta f(x;\\theta)\\) at current parameter \\(\\theta\\).
+        """
```
Comment (aleximmer) [inline laplace/curvature/curvature.py:52]: Comment should start right after `"""` as in all docstrings. What's the reason for the deviation here?

**Review by aleximmer — COMMENTED** [review]
Looks overall good to me. Please have a look at the two comments and let me know what you think. After that it's good to go from my side.

Thanks for taking on the effort and including curvlinops!

File: laplace/curvature/curvature.py [hunk 5]
```diff
@@ -188,38 +229,39 @@ class GGNInterface(CurvatureInterface):
...
-        loss : torch.Tensor
-        H_ggn : torch.Tensor
-            full GGN approximation `(parameters, parameters)`
-        """
-        loss = self.factor * self.lossfunc(f, y)
+            H_lik += 1/self.num_samples * torch.einsum('bc,bk->bck', grad_sample, grad_sample)
+
+        return H_lik
+
+    def _get_exact_H_lik(self, f):
```
Comment (aleximmer) [inline laplace/curvature/curvature.py:257]: Can we leave this is `_get_full_ggn` and add the function `_get_full_fisher` instead of these renamings? I `exact_H_lik` is inappropriate because it is not the true Hessian and would in principle have to be `log_lik`. `sample_H_lik` also sounds a bit different from what it actually is and just using `fisher` seems more appropriate to me.

**Review by aleximmer — COMMENTED** [review]
Sorry, forgot to add this comment in the previous review.

File: laplace/curvature/curvature.py (hunk 5 above)
Comment (wiseodd) [reply laplace/curvature/curvature.py]: I renamed them to better describe the functions. I think it's best to have them instead since there are many code overlaps in the `_get_full_ggn` and the proposed `_get_full_fisher` (also overlap with `diag_ggn` and `diag_fisher`).
